File: 310971/Guzik/task-320/src/main/java/com/example/entitiesapp/services/StickerService.java
package com.example.entitiesapp.services;

import com.example.entitiesapp.dto.request.StickerRequestTo;
import com.example.entitiesapp.dto.response.StickerResponseTo;
import com.example.entitiesapp.entities.Sticker;
import com.example.entitiesapp.exceptions.DuplicateResourceException;
import com.example.entitiesapp.exceptions.ResourceNotFoundException;
import com.example.entitiesapp.exceptions.ValidationException;
import com.example.entitiesapp.mappers.StickerMapper;
import com.example.entitiesapp.repositories.StickerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class StickerService {
    private final StickerRepository stickerRepository;
    private final StickerMapper stickerMapper;

    public List<StickerResponseTo> getAll() {
        return stickerRepository.findAll().stream()
                .map(stickerMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    public StickerResponseTo getById(Long id) {
        Sticker sticker = stickerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Sticker not found with id: " + id));
        return stickerMapper.toResponseDto(sticker);
    }

    @Transactional
    public StickerResponseTo create(StickerRequestTo dto) {
        validateStickerRequest(dto);

        if (stickerRepository.existsByName(dto.getName())) {
            throw new DuplicateResourceException("Sticker with name '" + dto.getName() + "' already exists");
        }

        Sticker sticker = stickerMapper.toEntity(dto);
        Sticker saved = stickerRepository.save(sticker);
        return stickerMapper.toResponseDto(saved);
    }

    @Transactional
    public StickerResponseTo update(Long id, StickerRequestTo dto) {
        validateStickerRequest(dto);

        Sticker existing = stickerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Sticker not found with id: " + id));

        if (!existing.getName().equals(dto.getName()) &&
                stickerRepository.existsByName(dto.getName())) {
            throw new DuplicateResourceException("Sticker with name '" + dto.getName() + "' already exists");
        }

        existing.setName(dto.getName());
        Sticker updated = stickerRepository.save(existing);
        return stickerMapper.toResponseDto(updated);
    }

    @Transactional
    public void delete(Long id) {
        if (!stickerRepository.existsById(id)) {
            throw new ResourceNotFoundException("Sticker not found with id: " + id);
        }
        stickerRepository.deleteById(id);
    }

    public List<StickerResponseTo> findByArticleId(Long articleId) {
        return stickerRepository.findByArticleId(articleId).stream()
                .map(stickerMapper::toResponseDto)
                .collect(Collectors.toList());
    }

    private void validateStickerRequest(StickerRequestTo dto) {
        if (dto.getName() == null || dto.getName().trim().isEmpty()) {
            throw new ValidationException("Name is required");
        }
        if (dto.getName().length() < 2 || dto.getName().length() > 32) {
            throw new ValidationException("Name must be between 2 and 32 characters");
        }
    }
}

File: 351004/Leshok/Lab6/services/StickerService.js
// services/StickerService.js
const stickerRepo = require('../repositories/StickerRepository');
const newsStickerRepo = require('../repositories/NewsStickerRepository');
const Sticker = require('../models/Sticker');
const { toResponse, fromRequest } = require('../dto/stickerDto');
const { ValidationError, NotFoundError } = require('../utils/errors');

class StickerService {
  async getAll() {
    return stickerRepo.findAll().map(toResponse);
  }

  async getById(id) {
    const sticker = stickerRepo.findById(id);
    if (!sticker) throw new NotFoundError('Sticker не найден');
    return toResponse(sticker);
  }

  async create(stickerRequest) {
    const { name } = fromRequest(stickerRequest);
    if (!name || name.trim() === '') {
      throw new ValidationError('Name не может быть пустым');
    }
    // Проверка уникальности имени (опционально)
    const existing = stickerRepo.findAll().find(s => s.name === name);
    if (existing) throw new ValidationError('Sticker с таким именем уже существует', 409);
    const sticker = new Sticker(null, name);
    const saved = stickerRepo.create(sticker);
    return toResponse(saved);
  }

  async update(id, stickerRequest) {
    const existing = stickerRepo.findById(id);
    if (!existing) throw new NotFoundError('Sticker не найден');
    const updateData = {};
    if (stickerRequest.name !== undefined) {
      if (stickerRequest.name.trim() === '') throw new ValidationError('Name не может быть пустым');
      updateData.name = stickerRequest.name;
    }
    const updated = stickerRepo.update(id, updateData);
    return toResponse(updated);
  }

  async delete(id) {
    const existing = stickerRepo.findById(id);
    if (!existing) throw new NotFoundError('Sticker не найден');
    // Удаляем связи с новостями
    const newsIds = newsStickerRepo.getNewsIdsByStickerId(id);
    for (let newsId of newsIds) {
      newsStickerRepo.removeStickerFromNews(newsId, id);
    }
    stickerRepo.delete(id);
    return true;
  }

  // Дополнительно: получить стикеры по id новости (для эндпоинта /news/:id/stickers)
  async getByNewsId(newsId) {
    const stickerIds = newsStickerRepo.getStickersByNewsId(newsId);
    const stickers = stickerIds.map(id => stickerRepo.findById(id)).filter(s => s);
    return stickers.map(toResponse);
  }
}

module.exports = new StickerService();